index.js

```diff
@@ -0,0 +1,979 @@
+          var result = {};
+          for (var k in this) result[k] = this[k];
+          for (k in strMap) result[k] = strMap[k];
+          return result;
+        },
+
+        //  Object#map :: StrMap a ~> (a -> b) -> StrMap b
+        'fantasy-land/map': function(f) {
+          var result = {};
+          for (var k in this) result[k] = f(this[k]);
+          return result;
+        },
+
+        //  Object#reduce :: StrMap a ~> ((b, a) -> b, b) -> b
+        'fantasy-land/reduce': function(f, initial) {
+          var result = initial;
+          for (var k in this) result = f(result, this[k]);
+          return result;
+        }
+      }
+    },
+
+    Function: {
+      prototype: {
+        //  Function#toString :: Function ~> () -> String
+        toString: function() { return String(this); },
+
+        //  Function#equals :: Function ~> Function -> Boolean
+        'fantasy-land/equals': function(other) { return other === this; },
+
+        //  Function#map :: (a -> b) ~> (b -> c) -> (a -> c)
+        'fantasy-land/map': function(f) {
+          var functor = this;
+          return function(x) { return f(functor(x)); };
+        },
+
+        //  Function#promap :: (b -> c) ~> (a -> b, c -> d) -> (a -> d)
+        'fantasy-land/promap': function(before, after) {
+          var profunctor = this;
+          return function(x) { return after(profunctor(before(x))); };
+        },
+
+        //  Function#ap :: (a -> b) ~> (a -> b -> c) -> (a -> c)
+        'fantasy-land/ap': function(a2b2c) {
+          var a2b = this;
+          return function(x) { return a2b2c(x)(a2b(x)); };
+        }
+      }
+    }
+
+  };
+
+  //  $seenStack :: Array (Array Any)
+  //
+  //  The strategy for handling circular references is as follows:
+  //
+  //  1.  ‘toString’ pushes a new empty array onto $seenStack.
+  //
+  //  2.  ‘toString’ calls the private ‘show’ function inside a ‘try’ block.
+  //
+  //  3.  ‘show’ binds the last (most recently added) element of $seenStack
+  //      to a local variable named $seen.
+  //
+  //  4.  ‘show’ determines whether the value to be shown appears in $seen,
+  //      using reference equality. If it is, a circular reference has been
+  //      detected, and ‘show’ returns '<Circular>'. Otherwise, it continues.
+  //
+  //  5.  ‘show’ pushes the value to be shown onto $seen.
+  //
+  //  6.  ‘show’ dispatches to a ‘toString’ method inside a ‘try’ block.
+  //
+  //  7.  The ‘finally’ block in ‘show’ pops the value pushed onto $seen in
+  //      step 5.
+  //
+  //  8.  If the method invocation in step 6 succeeded, its result is returned.
+  //      Otherwise, the exception is allowed to propagate.
+  //
+  //  9.  The ‘finally’ block in ‘toString’ pops the array pushed onto
+  //      $seenStack in step 1.
+  //
+  //  10. If the call to ‘show’ in step 2 succeeded, its result is returned.
+  //      Otherwise, the exception is allowed to propagate.
+  var $seenStack = [];
```

Comment:
I'd love someone (@Avaq?) to verify that this approach is sound. It's necessarily different from the approach used for [`R.toString`][1]. In Ramda's case there's a single recursive `toString` function with one case for each type, whereas in our case each type has its own `toString` method. Since we can't expect a `toString` method to accept an array of "seen" values as an argument, we instead rely on "global" mutable state. This should be okay given JavaScript's single-threaded execution model.

Check out the uses of the wonderfully complex and surprising [`try...finally`][2] statement. I don't think I had ever used it before today!


[1]: http://ramdajs.com/0.22.1/docs/#toString
[2]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/try...catch